docs/architecture/grpc-for-wcf-developers/apm.md

```diff
@@ -1,11 +1,278 @@
+}
+```
+
+#### Writing an event to a DiagnosticListener
+
+To avoid the overhead of writing diagnostics when no consumer is listening, you should always call the `IsEnabled` method on the listener before writing to it. If a consumer is listening, you can write an event providing a name and an object with information about the event. This can be any object: a gRPC request, a type from your own code, or even an anonymous object.
+
+```csharp
+public class StockDataService
+{
+    public override async Task<GetResponse> Get(GetRequest request, ServerCallContext context)
+    {
+        if (StockDataDiagnostics.Listener.IsEnabled("StockDataService.Get"))
+        {
+            StockDataDiagnostics.Listener.Write("StockDataService.Get", request);
+        }
+    }
+}
+```
+
+#### Tracing an Activity with a DiagnosticListener
+
+Again, you should check to see whether a listener is enabled before creating and starting an `Activity`. You do not need to write code to handle parent activities; this is handled internally using an [AsyncLocal](https://docs.microsoft.com/dotnet/api/system.threading.asynclocal-1?view=netcore-3.0) static property, `Current`, on the `Activity` class. Use the `StartActivity` and `StopActivity` methods on the listener to record the operation, including its duration.
+
+```csharp
+public class StockDataService
+{
+    public override async Task<GetResponse> Get(GetRequest request, ServerCallContext context)
+    {
+        Activity activity = null;
+        if (StockDataDiagnostics.Listener.IsEnabled("StockDataService.Get"))
+        {
+            activity = new Activity("StockDataService.Get");
+            StockDataDiagnostics.Listener.StartActivity(activity, request);
+        }
+        try
+        {
+            // Handle request
+        }
+        finally
+        {
+            if (activity != null)
+            {
+                StockDataDiagnostics.Listener.StopActivity(activity, request);
+            }
+        }
+    }
+}
+```
+
+The `Activity` class supports the tracing span concept of *baggage*: there is an `AddBaggage` method that takes a `string` key and a `string` value, and these values will be propagated by `HttpClient` and the gRPC client.
+
+```csharp
+var activity = new Activity("StockDataService.Get");
+activity.AddBaggage("InstanceId", _myInstanceId);
+```
+
+### Writing distributed trace data
+
+There is [an OpenTracing NuGet package](https://www.nuget.org/packages/OpenTracing/) which supports all OpenTracing-compliant back-ends. This package can be used independently of `DiagnosticSource`. But there is an additional package from the OpenTracing API Contributions project, [OpenTracing.Contrib.NetCore](https://www.nuget.org/packages/OpenTracing.Contrib.NetCore/), which adds a `DiagnosticSource` listener and writes events and activities to a back-end automatically. Enabling this package is as simple as installing it from NuGet and adding it as a service in your `Startup` class.
```

Comment:
New hotness: https://github.com/open-telemetry/opentelemetry-dotnet

Should use this rather than open tracing. Note: it's pretty new and rough

Reply:
It's incredibly new and rough. I'd prefer to leave the (slightly) more stable OpenTracing stuff in there for now and add a note that OpenTelemetry will be the preferred option in the future, maybe come back and update the book when it's stabilised.

Reply:
Added a note about OpenTelemetry .NET SDK but kept OpenTracing text with a caveat. https://github.com/dotnet/docs/pull/14438/files#diff-02130034a4b7526dbd006328cb690893R191

Reply:
I actually demo'd the OpenTelemetry stuff with Jaeger at a user group last night, and the alpha5 they've currently got on Nuget is completely busted. I had to use their Myget to get alpha26 which worked OK, but yeah: it's rough.